File: CodingTest/LeetCode/206-reverse-linked-list.js
/**
 * https://leetcode.com/submissions/detail/600161190/
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    let [prev, cur] = [null, head];
    while(cur) {
        const temp = prev;
        prev = cur;
        cur = cur.next;
        prev.next = temp;
    }
    return prev;
};

/**
 * https://leetcode.com/submissions/detail/600158316/
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    let [prev, cur] = [null, head];
    while(cur) {
        [cur.next, prev, cur] = [prev, cur, cur.next]
    }
    return prev;
};

File: CodingTest/LeetCode/206-reverse-linked-list.java
/**
 * https://leetcode.com/submissions/detail/623231081/
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode pre = null;
        ListNode curr = head;
        while(curr != null){
            ListNode temp = pre;
            pre = curr;
            curr = curr.next;
            pre.next = temp;
        }

        return pre;
    }
}